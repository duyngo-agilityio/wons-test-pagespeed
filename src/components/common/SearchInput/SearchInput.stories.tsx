import type { Meta, StoryObj } from '@storybook/react';

import { SearchInput } from '@/components';

const meta: Meta<typeof SearchInput> = {
  title: 'Components/Common/SearchInput',
  component: SearchInput,
  args: {},
  argTypes: {
    isDisabled: {
      description: 'Disable field',
      type: 'boolean',
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    className: {
      description: 'Change style as you like',
      type: 'string',
    },
    onChange: {
      description: 'Changes value of field',
      type: 'function',
    },
  },
};

export default meta;
type Story = StoryObj<typeof SearchInput>;

export const Default: Story = {
  args: {},
};

export const Disabled: Story = {
  args: {
    isDisabled: true,
  },
};

export const HasValue: Story = {
  args: {
    defaultValue: 'John Deo',
  },
};
