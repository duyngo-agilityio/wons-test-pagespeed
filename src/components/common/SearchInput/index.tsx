'use client';

import React, { memo } from 'react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { useDebouncedCallback } from 'use-debounce';
import { Input, InputProps } from '@nextui-org/react';

// Constants
import { SEARCH_QUERIES, DEFAULT_PAGE } from '@/constants';

// Components
import { CiSearch } from '@/components';

const SearchInput = ({ ...props }: InputProps) => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();
  const defaultValueSearch: string = searchParams
    ?.get(SEARCH_QUERIES.QUERY)
    ?.toString() as string;

  const { PAGE, QUERY } = SEARCH_QUERIES;

  const handleSearch = useDebouncedCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const term = event.target.value;
      const params = new URLSearchParams(searchParams);

      params.set(PAGE, DEFAULT_PAGE.toString());

      if (term) {
        params.set(QUERY, term);
      } else {
        params.delete(QUERY);
      }

      replace(`${pathname}?${params.toString()}`);
    },
    500,
  );

  return (
    <Input
      classNames={{
        input: [
          '!text-blue-800 dark:!text-white',
          'placeholder:text-sm placeholder:text-blue-800/50 dark:placeholder:text-white/50',
        ],
        inputWrapper: [
          'bg-white dark:bg-gray-400',
          'hover:!bg-white dark:hover:!bg-gray-400',
          'focus-within:!bg-white dark:focus-within:!bg-gray-400',
          'group-data-[focus=true]:!bg-white dark:group-data-[focus=true]:!bg-gray-400',
          'pl-5 pr-3.5 py-3',
          'md:max-w-[230px]',
        ],
      }}
      placeholder="Search"
      defaultValue={defaultValueSearch}
      onChange={handleSearch}
      endContent={
        <CiSearch className="w-3 h-3 text-blue-800/60 dark:text-white/60 cursor-pointer" />
      }
      {...props}
    />
  );
};

export default memo(SearchInput);
