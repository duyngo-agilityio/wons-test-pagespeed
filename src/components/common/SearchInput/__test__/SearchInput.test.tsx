// Components
import SearchInput from '../index';

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useRouter: jest.fn().mockImplementation(() => ({
    replace: jest.fn(),
  })),
  useSearchParams: jest.fn(() => ({
    get: () => () => 'John Deo',
  })),
  usePathname: jest.fn(),
}));

describe('SearchInput', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('should match snapshot', () => {
    const { container } = testLibJestUtils.render(<SearchInput />);

    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it('calls change values', async () => {
    const { container, getByPlaceholderText } = testLibJestUtils.render(
      <SearchInput />,
    );

    testLibJestUtils.fireEvent.change(getByPlaceholderText('Search'), {
      target: {
        value: 'John Deo',
      },
    });

    // Wait for it to add the input value
    jest.advanceTimersByTime(2000);

    expect(container).toMatchSnapshot();
  });

  it('changes empty value', async () => {
    const { container, getByPlaceholderText } = testLibJestUtils.render(
      <SearchInput />,
    );

    testLibJestUtils.fireEvent.change(getByPlaceholderText('Search'), {
      target: {
        value: 'John Deo',
      },
    });
    testLibJestUtils.fireEvent.change(getByPlaceholderText('Search'), {
      target: {
        value: '',
      },
    });

    // Wait for it to add the input value
    jest.advanceTimersByTime(2000);

    expect(container).toMatchSnapshot();
  });
});
