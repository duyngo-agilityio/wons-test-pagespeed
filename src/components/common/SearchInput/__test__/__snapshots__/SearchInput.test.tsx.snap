// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput calls change values 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
     
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)] is-filled"
      data-filled="true"
      data-filled-within="true"
      data-has-value="true"
      data-slot="base"
    >
      <div
        class="h-full flex flex-col"
        data-slot="main-wrapper"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background pl-5 pr-3.5 py-3 max-w-[230px] is-filled"
          data-slot="input-wrapper"
          style="cursor: text;"
        >
          <div
            class="inline-flex w-full items-center h-full box-border"
            data-slot="inner-wrapper"
          >
            <input
              aria-label="Search"
              class="w-full font-normal !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground bg-white dark:bg-gray-400 !text-blue-800/70 placeholder:text-sm placeholder:text-blue-800/50 is-filled"
              data-filled="true"
              data-filled-within="true"
              data-has-end-content="true"
              data-slot="input"
              id="react-aria-:r5:"
              placeholder="Search"
              type="text"
              value="John Deo"
            />
            <svg
              class="w-3 h-3 text-blue-800/60"
              fill="currentColor"
              height="1em"
              id="search"
              stroke="currentColor"
              stroke-width="0"
              version="1.1"
              viewBox="0 0 24 24"
              width="1em"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <g>
                <path
                  d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed top-4 right-4"
    />
  </div>
</div>
`;

exports[`SearchInput changes empty value 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
     
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
      data-filled="true"
      data-filled-within="true"
      data-slot="base"
    >
      <div
        class="h-full flex flex-col"
        data-slot="main-wrapper"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background pl-5 pr-3.5 py-3 max-w-[230px]"
          data-slot="input-wrapper"
          style="cursor: text;"
        >
          <div
            class="inline-flex w-full items-center h-full box-border"
            data-slot="inner-wrapper"
          >
            <input
              aria-label="Search"
              class="w-full font-normal !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground bg-white dark:bg-gray-400 !text-blue-800/70 placeholder:text-sm placeholder:text-blue-800/50"
              data-has-end-content="true"
              data-slot="input"
              id="react-aria-:ra:"
              placeholder="Search"
              type="text"
              value=""
            />
            <svg
              class="w-3 h-3 text-blue-800/60"
              fill="currentColor"
              height="1em"
              id="search"
              stroke="currentColor"
              stroke-width="0"
              version="1.1"
              viewBox="0 0 24 24"
              width="1em"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <g>
                <path
                  d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed top-4 right-4"
    />
  </div>
</div>
`;

exports[`SearchInput should match snapshot 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
     
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
      data-filled="true"
      data-filled-within="true"
      data-slot="base"
    >
      <div
        class="h-full flex flex-col"
        data-slot="main-wrapper"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background pl-5 pr-3.5 py-3 max-w-[230px]"
          data-slot="input-wrapper"
          style="cursor: text;"
        >
          <div
            class="inline-flex w-full items-center h-full box-border"
            data-slot="inner-wrapper"
          >
            <input
              aria-label="Search"
              class="w-full font-normal !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground bg-white dark:bg-gray-400 !text-blue-800/70 placeholder:text-sm placeholder:text-blue-800/50"
              data-has-end-content="true"
              data-slot="input"
              id="react-aria-:r0:"
              placeholder="Search"
              type="text"
              value=""
            />
            <svg
              class="w-3 h-3 text-blue-800/60"
              fill="currentColor"
              height="1em"
              id="search"
              stroke="currentColor"
              stroke-width="0"
              version="1.1"
              viewBox="0 0 24 24"
              width="1em"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <g>
                <path
                  d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed top-4 right-4"
    />
  </div>
</div>
`;
